sycl/test/extensions/properties/new_properties_open_question.cpp

```diff
@@ -0,0 +1,185 @@
+// RUN: %clangxx -fsycl -fsyntax-only %s
+
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+using namespace sycl::ext::oneapi::experimental::new_properties;
+
+template <typename property_key_t> constexpr auto generate_property_key_name() {
+#if defined(__clang__) || defined(__GNUC__)
+  return __PRETTY_FUNCTION__;
+#elif defined(_MSC_VER)
+  return __FUNCSIG__;
+#else
+#error "Unsupported compiler"
+#endif
+}
+
+template <typename property_t, typename property_key_t = property_t>
+struct named_property_base
+    : public detail::property_base<property_t, property_key_t> {
+  static constexpr std::string_view property_name =
+      generate_property_key_name<property_key_t>();
+};
+
+struct prop_key {};
+
+template <typename PropList> auto foo(PropList props = {}) {
+  static_assert(props.template has_property<prop_key>());
+  static_assert(props.template get_property<prop_key>().value == 42);
+}
+
+namespace approach_one {
+// Don't provide `inline constexpr` property variables at all, they don't
+// provide enough benefits now that `property_value` thing was eliminated.
+
+template <int N> struct prop : named_property_base<prop<N>, prop_key> {
+  static constexpr int value = N;
+};
+struct other_prop : named_property_base<other_prop> {};
+
+void bar() {
+  foo(properties{prop<42>{}});
+  foo(properties{prop<42>{}, other_prop{}});
```

Comment:
I like the simplicity of this.  I think the old approach required us to define too many things with the same/similar names, and so if we can avoid having to define the shortcuts that's probably better for everybody.

Another benefit of exposing that these property values are objects is that developers will naturally have a better understanding of the difference between properties with compile-time and run-time values:
- `prop<42>{}` is clearly an instance of a class storing 42 as a compile-time value
- `prop(42)` is clearly an instance of a class storing 42 as a run-time value

Reply:
I don't agree.  I think it's still clear that `prop<42>` is a compile-time property (without the `{}`), and I like that it's two characters shorter.

Also, I think the inline constexpr values are a verbosity saver for compile-time properties that have an enumeration.  Consider the `host_acess` property from [sycl_ext_oneapi_device_global](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc):

```
properties props{host_access_read};                       // with inline constexpr variable
properties props{host_access<host_access_enum::read>{}};  // without
```

Reply:
> It's shorter, but it's inconsistent. In the situation where `prop` supports both compile-time and run-time values, the compile-time value would be used as `prop<42>` and the run-time value would be used as `prop_property(42)`. Pairing `prop<42>{}` and `prop(42)` is more consistent, and we don't require developers to learn any additional naming conventions.

We had this discussion before. There are no properties today that mix compile-time and runtime values, and we don't have any immediate plans to add any like this.  I do not think it makes sense to optimize for such a weird case, which will probably never happen.

Reply:
FYI `work_group_scratchpad_size` is currently implemented as a compile-time/runtime property, although it is not meant to have the payload statically set but only to trigger compiler transformations.

Reply:
I don't have much of a perspective on this, but if it is desirable to allow a single property to support a value being provided as either a compile-time or run-time value, that can be done (and perhaps it already is, I wouldn't know); https://godbolt.org/z/qcTzKG8jY. I haven't been able to figure out a way to avoid requiring the `{}` initializer for the compile-time case though.

Reply:
I'd envision something like this instead: https://godbolt.org/z/b97a39zrv

Reply:
I was trying to avoid stealing a value from the representable range of value. But yes, that works too.

You can avoid the `<>` by employing a deduction guide. https://godbolt.org/z/zK53Kvvnv.

Reply:
https://godbolt.org/z/PYs946Ko4 avoids stealing.

Reply:
I agree that's clever, but it's still not clear to me that there's a real use case here.  Why would we want a compile-time-constant version of `work_group_scratchpad_size`?  Is there some meaningful optimization that we expect the compiler to do in this case?

In any case, this would be a very disruptive change in the common use case for `properties`.  There's a lot of code out there using the inline constexpr variables.  We can't ask all that code to change now.